packages/core/src/cli/utils/patch-program.ts

```diff
@@ -0,0 +1,45 @@
+import type TS from 'typescript';
+import type TransformManager from '../../common/transform-manager';
+import { assert } from './assert';
+import type TransformManagerPool from './transform-manager-pool';
+
+type Program = TS.SemanticDiagnosticsBuilderProgram;
+
+export function patchProgram(
+  program: Program,
+  transformManagerOrPool: TransformManagerPool | TransformManager
+): void {
+  let { getSyntacticDiagnostics, getSemanticDiagnostics } = program;
+
+  // We have two scenarios: either we are working with a single program for a
+  // single project, in which case we have a `TransformManager`, *or* we have a
+  // `TransformManagerPool` to deal with the case where we are running a
+  // `--build` across a composite project. In the latter case, we need to *get*
+  // a manager for the project in question, if it exists at all, to use in
+  // getting transform errors (both type errors and syntax errors).
+  let manager: TransformManager | null;
+  if (isPool(transformManagerOrPool)) {
+    let configFile = program.getCompilerOptions()['configFilePath'];
+    assert(typeof configFile === 'string', 'internal error: missing TS config file');
+    manager = transformManagerOrPool.managerFor(configFile);
+  } else {
+    manager = transformManagerOrPool;
+  }
```

Comment:
I know these 8 lines were the result of a lot of blood, sweat and tears, but I'm pretty happy with how this all came together!